File: code/programs/rust/unix-tools/tests/xargs_test.rs
//! # Integration Tests for xargs
//!
//! These tests verify the xargs pipeline: input splitting, item
//! grouping, command building, and execution.

use unix_tools::xargs_tool::*;

// ---------------------------------------------------------------------------
// Tests: Input splitting
// ---------------------------------------------------------------------------

#[cfg(test)]
mod splitting {
    use super::*;

    #[test]
    fn split_whitespace() {
        let items = split_input("a b c", &XargsOptions::default());
        assert_eq!(items, vec!["a", "b", "c"]);
    }

    #[test]
    fn split_newlines() {
        let items = split_input("a\nb\nc", &XargsOptions::default());
        assert_eq!(items, vec!["a", "b", "c"]);
    }

    #[test]
    fn split_tabs() {
        let items = split_input("a\tb\tc", &XargsOptions::default());
        assert_eq!(items, vec!["a", "b", "c"]);
    }

    #[test]
    fn split_mixed_whitespace() {
        let items = split_input("a  b\t\tc\n\nd", &XargsOptions::default());
        assert_eq!(items, vec!["a", "b", "c", "d"]);
    }

    #[test]
    fn split_null_delimited() {
        let opts = XargsOptions { null_delim: true, ..Default::default() };
        let items = split_input("a\0b\0c\0", &opts);
        assert_eq!(items, vec!["a", "b", "c"]);
    }

    #[test]
    fn split_null_preserves_spaces() {
        let opts = XargsOptions { null_delim: true, ..Default::default() };
        let items = split_input("file with spaces\0another\0", &opts);
        assert_eq!(items, vec!["file with spaces", "another"]);
    }

    #[test]
    fn split_custom_delimiter() {
        let opts = XargsOptions { delimiter: Some(':'), ..Default::default() };
        let items = split_input("/usr/bin:/bin:/sbin", &opts);
        assert_eq!(items, vec!["/usr/bin", "/bin", "/sbin"]);
    }

    #[test]
    fn split_empty_input() {
        let items = split_input("", &XargsOptions::default());
        assert!(items.is_empty());
    }
}

// ---------------------------------------------------------------------------
// Tests: Item grouping
// ---------------------------------------------------------------------------

#[cfg(test)]
mod grouping {
    use super::*;

    #[test]
    fn group_all_in_one() {
        let items = vec!["a".into(), "b".into(), "c".into()];
        let groups = group_items(items, &XargsOptions::default());
        assert_eq!(groups.len(), 1);
    }

    #[test]
    fn group_by_two() {
        let items = vec!["a".into(), "b".into(), "c".into(), "d".into(), "e".into()];
        let opts = XargsOptions { max_args: Some(2), ..Default::default() };
        let groups = group_items(items, &opts);
        assert_eq!(groups.len(), 3);
        assert_eq!(groups[0].len(), 2);
        assert_eq!(groups[1].len(), 2);
        assert_eq!(groups[2].len(), 1);
    }

    #[test]
    fn group_by_one() {
        let items = vec!["a".into(), "b".into(), "c".into()];
        let opts = XargsOptions { max_args: Some(1), ..Default::default() };
        let groups = group_items(items, &opts);
        assert_eq!(groups.len(), 3);
    }

    #[test]
    fn group_empty() {
        let items: Vec<String> = vec![];
        let groups = group_items(items, &XargsOptions::default());
        assert!(groups.is_empty());
    }
}

// ---------------------------------------------------------------------------
// Tests: Command building
// ---------------------------------------------------------------------------

#[cfg(test)]
mod building {
    use super::*;

    #[test]
    fn build_append() {
        let cmd = vec!["echo".into()];
        let items = vec!["a".into(), "b".into()];
        let commands = build_command_args(&cmd, &items, &None);
        assert_eq!(commands, vec![vec!["echo", "a", "b"]]);
    }

    #[test]
    fn build_replace() {
        let cmd = vec!["echo".into(), "{}".into()];
        let items = vec!["hello".into(), "world".into()];
        let replace = Some("{}".to_string());
        let commands = build_command_args(&cmd, &items, &replace);
        assert_eq!(commands.len(), 2);
        assert_eq!(commands[0], vec!["echo", "hello"]);
        assert_eq!(commands[1], vec!["echo", "world"]);
    }

    #[test]
    fn build_replace_in_multiple_args() {
        let cmd = vec!["mv".into(), "{}".into(), "{}.bak".into()];
        let items = vec!["file.txt".into()];
        let replace = Some("{}".to_string());
        let commands = build_command_args(&cmd, &items, &replace);
        assert_eq!(commands[0], vec!["mv", "file.txt", "file.txt.bak"]);
    }

    #[test]
    fn build_with_no_items() {
        let cmd = vec!["echo".into()];
        let items: Vec<String> = vec![];
        let commands = build_command_args(&cmd, &items, &None);
        // Should just be the command with no extra args
        assert_eq!(commands, vec![vec!["echo"]]);
    }
}

// ---------------------------------------------------------------------------
// Tests: Execution
// ---------------------------------------------------------------------------

#[cfg(test)]
mod execution {
    use super::*;

    #[test]
    fn execute_echo_succeeds() {
        let args = vec!["echo".into(), "test".into()];
        let result = execute_command(&args, false);
        assert_eq!(result.unwrap(), 0);
    }

    #[test]
    fn execute_false_returns_nonzero() {
        let args = vec!["false".into()];
        let result = execute_command(&args, false);
        assert_ne!(result.unwrap(), 0);
    }

    #[test]
    fn execute_nonexistent_fails() {
        let args = vec!["not_a_command_xyzzy_123".into()];
        let result = execute_command(&args, false);
        assert!(result.is_err());
    }

    #[test]
    fn execute_empty_args_fails() {
        let args: Vec<String> = vec![];
        let result = execute_command(&args, false);
        assert!(result.is_err());
    }
}

// ---------------------------------------------------------------------------
// Tests: Full pipeline
// ---------------------------------------------------------------------------

#[cfg(test)]
mod pipeline {
    use super::*;

    #[test]
    fn run_xargs_echo() {
        let result = run_xargs("a b c", &["echo".into()], &XargsOptions::default());
        assert_eq!(result.unwrap(), 0);
    }

    #[test]
    fn run_xargs_no_run_if_empty() {
        let opts = XargsOptions { no_run_if_empty: true, ..Default::default() };
        let result = run_xargs("", &["echo".into()], &opts);
        assert_eq!(result.unwrap(), 0);
    }

    #[test]
    fn run_xargs_with_max_args() {
        let opts = XargsOptions { max_args: Some(1), ..Default::default() };
        let result = run_xargs("a b c", &["echo".into()], &opts);
        assert_eq!(result.unwrap(), 0);
    }

    #[test]
    fn run_xargs_with_replace() {
        let opts = XargsOptions {
            replace_str: Some("{}".to_string()),
            max_args: Some(1),
            ..Default::default()
        };
        let result = run_xargs("hello world", &["echo".into(), "item: {}".into()], &opts);
        assert_eq!(result.unwrap(), 0);
    }
}

File: code/programs/typescript/unix-tools/tests/xargs.test.ts
/**
 * Tests for xargs -- build and execute command lines from standard input.
 *
 * We test the exported business logic functions: parseItems,
 * buildBatches, buildCommands, and runXargs.
 */

import { describe, it, expect } from "vitest";
import {
  parseItems,
  buildBatches,
  buildCommands,
  runXargs,
  XargsOptions,
} from "../src/xargs.js";

// ---------------------------------------------------------------------------
// Helper: default xargs options (no flags set).
// ---------------------------------------------------------------------------

function defaultOpts(overrides: Partial<XargsOptions> = {}): XargsOptions {
  return {
    nullDelimiter: false,
    delimiter: null,
    maxArgs: 0,
    replaceStr: null,
    verbose: false,
    noRunIfEmpty: false,
    ...overrides,
  };
}

// ---------------------------------------------------------------------------
// parseItems: input parsing tests.
// ---------------------------------------------------------------------------

describe("parseItems", () => {
  it("should split on whitespace by default", () => {
    const items = parseItems("hello world foo", defaultOpts());
    expect(items).toEqual(["hello", "world", "foo"]);
  });

  it("should handle multiple whitespace types", () => {
    const items = parseItems("hello\tworld\nfoo", defaultOpts());
    expect(items).toEqual(["hello", "world", "foo"]);
  });

  it("should handle leading and trailing whitespace", () => {
    const items = parseItems("  hello world  ", defaultOpts());
    expect(items).toEqual(["hello", "world"]);
  });

  it("should handle empty input", () => {
    const items = parseItems("", defaultOpts());
    expect(items).toEqual([]);
  });

  it("should handle whitespace-only input", () => {
    const items = parseItems("   \t\n  ", defaultOpts());
    expect(items).toEqual([]);
  });

  it("should respect single quotes", () => {
    const items = parseItems("'hello world' foo", defaultOpts());
    expect(items).toEqual(["hello world", "foo"]);
  });

  it("should respect double quotes", () => {
    const items = parseItems('"hello world" foo', defaultOpts());
    expect(items).toEqual(["hello world", "foo"]);
  });

  it("should handle backslash escaping", () => {
    const items = parseItems("hello\\ world foo", defaultOpts());
    expect(items).toEqual(["hello world", "foo"]);
  });

  it("should handle mixed quotes", () => {
    const items = parseItems("'single' \"double\" plain", defaultOpts());
    expect(items).toEqual(["single", "double", "plain"]);
  });

  // --- NUL delimiter mode (-0) -----------------------------------------

  it("should split on NUL in null mode", () => {
    const items = parseItems("hello\0world\0foo", defaultOpts({ nullDelimiter: true }));
    expect(items).toEqual(["hello", "world", "foo"]);
  });

  it("should handle trailing NUL", () => {
    const items = parseItems("hello\0world\0", defaultOpts({ nullDelimiter: true }));
    expect(items).toEqual(["hello", "world"]);
  });

  it("should preserve spaces in NUL mode", () => {
    const items = parseItems("hello world\0foo bar\0", defaultOpts({ nullDelimiter: true }));
    expect(items).toEqual(["hello world", "foo bar"]);
  });

  // --- Custom delimiter (-d) -------------------------------------------

  it("should split on custom delimiter", () => {
    const items = parseItems("hello,world,foo", defaultOpts({ delimiter: "," }));
    expect(items).toEqual(["hello", "world", "foo"]);
  });

  it("should handle custom delimiter with empty items", () => {
    const items = parseItems("hello,,foo", defaultOpts({ delimiter: "," }));
    expect(items).toEqual(["hello", "foo"]);
  });

  it("should split on newline as custom delimiter", () => {
    const items = parseItems("hello\nworld\nfoo", defaultOpts({ delimiter: "\n" }));
    expect(items).toEqual(["hello", "world", "foo"]);
  });
});

// ---------------------------------------------------------------------------
// buildBatches: batching items for execution.
// ---------------------------------------------------------------------------

describe("buildBatches", () => {
  it("should put all items in one batch when maxArgs is 0", () => {
    const batches = buildBatches(["a", "b", "c", "d"], 0);
    expect(batches).toEqual([["a", "b", "c", "d"]]);
  });

  it("should split items into batches of maxArgs", () => {
    const batches = buildBatches(["a", "b", "c", "d"], 2);
    expect(batches).toEqual([["a", "b"], ["c", "d"]]);
  });

  it("should handle uneven splits", () => {
    const batches = buildBatches(["a", "b", "c", "d", "e"], 2);
    expect(batches).toEqual([["a", "b"], ["c", "d"], ["e"]]);
  });

  it("should handle maxArgs larger than item count", () => {
    const batches = buildBatches(["a", "b"], 10);
    expect(batches).toEqual([["a", "b"]]);
  });

  it("should handle single item per batch", () => {
    const batches = buildBatches(["a", "b", "c"], 1);
    expect(batches).toEqual([["a"], ["b"], ["c"]]);
  });

  it("should return empty array for empty input", () => {
    const batches = buildBatches([], 2);
    expect(batches).toEqual([]);
  });

  it("should return empty array for empty input with maxArgs 0", () => {
    const batches = buildBatches([], 0);
    expect(batches).toEqual([]);
  });
});

// ---------------------------------------------------------------------------
// buildCommands: building command strings.
// ---------------------------------------------------------------------------

describe("buildCommands", () => {
  it("should append items to base command", () => {
    const cmds = buildCommands(["echo"], ["hello", "world"], null);
    expect(cmds).toEqual(["echo hello world"]);
  });

  it("should handle empty items", () => {
    const cmds = buildCommands(["echo"], [], null);
    expect(cmds).toEqual(["echo"]);
  });

  it("should handle command with initial args", () => {
    const cmds = buildCommands(["grep", "-l"], ["pattern", "file"], null);
    expect(cmds).toEqual(["grep -l pattern file"]);
  });

  it("should quote items with spaces", () => {
    const cmds = buildCommands(["echo"], ["hello world", "foo"], null);
    expect(cmds).toEqual(['echo "hello world" foo']);
  });

  // --- Replacement mode (-I) -------------------------------------------

  it("should replace placeholder with each item", () => {
    const cmds = buildCommands(["echo", "item: {}"], ["foo", "bar"], "{}");
    expect(cmds).toHaveLength(2);
    expect(cmds[0]).toContain("foo");
    expect(cmds[1]).toContain("bar");
  });

  it("should replace all occurrences of placeholder", () => {
    const cmds = buildCommands(["cp", "{}", "/backup/{}"], ["file.txt"], "{}");
    expect(cmds).toHaveLength(1);
    expect(cmds[0]).toContain("file.txt");
  });
});

// ---------------------------------------------------------------------------
// runXargs: end-to-end pipeline tests.
// ---------------------------------------------------------------------------

describe("runXargs", () => {
  it("should execute echo with items", () => {
    const result = runXargs("hello world", ["echo"], defaultOpts());
    expect(result.exitCode).toBe(0);
    expect(result.output.trim()).toBe("hello world");
  });

  it("should handle empty input with noRunIfEmpty", () => {
    const result = runXargs("", ["echo"], defaultOpts({ noRunIfEmpty: true }));
    expect(result.exitCode).toBe(0);
    expect(result.output).toBe("");
  });

  it("should batch with maxArgs", () => {
    const result = runXargs("a b c d", ["echo"], defaultOpts({ maxArgs: 2 }));
    expect(result.exitCode).toBe(0);
    // Should have two lines: "a b" and "c d".
    const lines = result.output.trim().split("\n");
    expect(lines).toHaveLength(2);
    expect(lines[0]).toBe("a b");
    expect(lines[1]).toBe("c d");
  });

  it("should handle replacement mode", () => {
    const result = runXargs(
      "foo\nbar",
      ["echo", "item: {}"],
      defaultOpts({ replaceStr: "{}" })
    );
    expect(result.exitCode).toBe(0);
    const lines = result.output.trim().split("\n");
    expect(lines).toHaveLength(2);
  });

  it("should handle NUL-delimited input", () => {
    const result = runXargs(
      "hello\0world",
      ["echo"],
      defaultOpts({ nullDelimiter: true })
    );
    expect(result.exitCode).toBe(0);
    expect(result.output.trim()).toBe("hello world");
  });

  it("should handle custom delimiter", () => {
    const result = runXargs(
      "hello,world",
      ["echo"],
      defaultOpts({ delimiter: "," })
    );
    expect(result.exitCode).toBe(0);
    expect(result.output.trim()).toBe("hello world");
  });

  it("should return non-zero exit code for failed commands", () => {
    const result = runXargs(
      "nonexistent_file",
      ["cat"],
      defaultOpts()
    );
    expect(result.exitCode).not.toBe(0);
  });

  it("should handle empty input with replacement mode", () => {
    const result = runXargs(
      "",
      ["echo", "{}"],
      defaultOpts({ replaceStr: "{}" })
    );
    expect(result.exitCode).toBe(0);
    expect(result.output).toBe("");
  });
});
